### PR TITLE
[VectorDistribution] Add distribution for trivial vector.extract

### PR DESCRIPTION
**Pull request by Groverkss**
This patch adds a distribution pattern for vector.extract when the list of indices is zero. This arises in the case of a scalar extract for 0-d vectors.

### CODE REVIEW

**Review by raikonenfnu — COMMENTED** [review]
IIUC, we want to distribute rank 0 vector extract to potentially different ranks vector extract depending on the layout. If so, can we get a lit test that showcases that?

**Comment by Groverkss** [comment]
> IIUC, we want to distribute rank 0 vector extract to potentially different ranks vector extract depending on the layout. If so, can we get a lit test that showcases that?

I think rank 0 vector.extract just folds away if the result isn't scalar. I can try adding one.

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributionPatterns.cpp [hunk 1]
```diff
@@ -313,10 +313,39 @@ struct DistributeGather final : OpDistributionPattern<vector::GatherOp> {
...
+    }
+
+    VectorValue source = extractOp.getVector();
+    VectorLayoutInterface sourceLayout = signature[source];
+
+    int64_t rank = sourceLayout.getRank();
+
+    Value distributed = rewriter.create<vector::ExtractOp>(
+        extractOp.getLoc(), getDistributed(rewriter, source, sourceLayout),
+        SmallVector<int64_t>(rank, 0));
```
Comment (manupak) [inline compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributionPatterns.cpp]: Shouldn;t this be an empty vector?

**Review by manupak — COMMENTED** [review]
highlevel question : how does this even get a layout propagated into it ?

File: compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributionPatterns.cpp (hunk 1 above)
Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributionPatterns.cpp]: even better: an empty array ref

Comment (Groverkss) [reply compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributionPatterns.cpp]: done

**Comment by Groverkss** [comment]
> > IIUC, we want to distribute rank 0 vector extract to potentially different ranks vector extract depending on the layout. If so, can we get a lit test that showcases that?
> 
> I think rank 0 vector.extract just folds away if the result isn't scalar. I can try adding one.

Actually, you are right. I updated the logic to only extract from rank 0 source vectors, which means the layout is always non distributed.

**Comment by Groverkss** [comment]
> highlevel question : how does this even get a layout propagated into it ?

Something like this: https://github.com/iree-org/iree/pull/19367 or through a reduction like vector.multi_reduction / vector.contract .

**Review by manupak — APPROVED** [review]
Whilst this LGTM, I cant help but wonder why we we are enforcing the layout for 0 d cases -- layout analysis could just stop there and not expose these to vector distribution.

**Pull request merged**